### ISSUE TITLE
SSL authentication fails against OpenShift 3.9

### ISSUE THREAD

**Issue by martinpitt** [cockpit-project/cockpit#8638]
I'm building an "openshift-prerelease" image to test against the next OpenShift version. This currently fails all tests as SSL certificate based authentication fails from Cockpit. ~/.kube/config (particularly its certificates) are valid, as on the command line `oc whoami` (and other commands) work fine.

This can be reproduced outside of the Cockpit UI with just cockpit-bridge. Extract the cert, key, and CA from ~/.kube/config into standard PEM files (they are just base64 encoded), and try to connect to the test VM with `cockpit-bridge --interact=---`:
```

{ "command": "open", "channel": "o", "payload": "http-stream2", "address": "10.111.112.101", "port": 8443, "tls": { "certificate": { "file": "/home/martin/prerel.pem" }, "key": { "file": "/home/martin/prerel.key" }, "authority": { "file": "/home/martin/prerel-ca.pem" }, "validate": true }, "method": "GET", "path": "/api/" }
---

{"command": "done", "channel": "o"}
---
```
This fails with
```
{"status":401,"reason":"Unauthorized","headers":{"Date":"Fri, 16 Feb 2018 14:37:48 GMT","Content-Type":"application/json","Cache-Control":"no-store"},"command":"response","channel":"o"}
---
o
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Unauthorized","reason":"Unauthorized","code":401}
---
```
The same command against current OpenShift 3.7 works fine:
```
{"status":200,"reason":"OK","headers":{"Date":"Fri, 16 Feb 2018 14:31:06 GMT","Content-Type":"application/json","Cache-Control":"no-store"},"command":"response","channel":"o"}
---
o
{"kind":"APIVersions","versions":["v1"],"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":"10.111.112.101:8443"}]}
```
The extracted certificates are also correct, as using curl or openssl works well:
```
openssl s_client -CAfile ~/prerel-ca.pem -connect 10.111.112.101:8443 -cert ~/prerel.pem
curl --cert ~/prerel.pem --cacert ~/prerel-ca.pem https://10.111.112.101:8443/api/v1/namespaces/default/pods
```

**Comment by martinpitt** [comment]
When that happens, the server logs this error:
```
openshift-run[847]: E0216 09:40:18.428176     854 authentication.go:64] Unable to authenticate the request due to an error: [x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "openshift-signer@1518732226"), x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "openshift-signer@1518732226")]
```

**Comment by martinpitt** [comment]
This is a convenient reproducer shell script which fetches the config from the running openshift (or openshift-prerelease) VM, extracts the keys, and calls cockpit-bridge for connecting to it.

```sh
#!/bin/sh
set -eu

# fetch config from `test/vm-run openshift`
cfg=$(ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o CheckHostIP=no -p 2201 root@127.0.0.2 cat .kube/config)

# extract certs
echo "$cfg" | awk '/client-certificate-data:/ { print $2 }' | base64 -d > /tmp/cert.pem
echo "$cfg" | awk '/client-key-data:/ { print $2 }' | base64 -d > /tmp/cert.key
echo "$cfg" | awk '/certificate-authority-data:/ { print $2 }' | base64 -d > /tmp/ca.pem


( sleep 0.5;
  printf '\n{ "command": "open", "channel": "o", "payload": "http-stream2", "address": "10.111.112.101", "port": 8443, "tls": { "certificate": { "file": "/tmp/cert.pem" }, "key": { "file": "/tmp/cert.key" }, "authority": { "file": "/tmp/ca.pem" }, "validate": true }, "method": "GET", "path": "/api/" }\n---\n';
  sleep 1
  printf '\n{"command": "done", "channel": "o"}\n---\n';
  sleep 1
) | cockpit-bridge --interact=---
```

**Comment by martinpitt** [comment]
Bah, I was wasting time on debuggging glib-networking, but it turns out that it's reproducible easily with gnutls-bin even:
```
❱❱❱ (printf 'GET /api HTTP/1.0\n\n'; sleep 1)  | gnutls-cli --x509certfile=/tmp/cert.pem --x509keyfile=/tmp/cert.key --x509cafile=/tmp/ca.pem -p 8443 10.111.112.101
Processed 1 CA certificate(s).
Processed 1 client X.509 certificates...
Resolving '10.111.112.101:8443'...
Connecting to '10.111.112.101:8443'...
- Certificate type: X.509
- Got a certificate list of 2 certificates.
- Certificate[0] info:
 - subject `CN=10.111.112.101', issuer `CN=openshift-signer@1518732226', serial 0x02, RSA key 2048 bits, signed using RSA-SHA256, activated `2018-02-15 22:03:46 UTC', expires `2020-02-15 22:03:47 UTC', pin-sha256="8s1sXBqKC8Vvwk/Mj/UawpbMS83ZFJDCiaKTZTEwK2w="
	Public Key ID:
		sha1:f4ed7eaacc40a78d09f4a20acac50c326d38ba46
		sha256:f2cd6c5c1a8a0bc56fc24fcc8ff51ac296cc4bcdd91490c289a2936531302b6c
	Public Key PIN:
		pin-sha256:8s1sXBqKC8Vvwk/Mj/UawpbMS83ZFJDCiaKTZTEwK2w=
	Public key's random art:
		+--[ RSA 2048]----+
		|                 |
		|                 |
		|      . .        |
		| o   . o . .     |
		|=.o   o S o .    |
		|oE+  . + * .     |
		|+  +.   = . .    |
		|o+..     + .  .  |
		|+..       +.oo   |
		+-----------------+

- Certificate[1] info:
 - subject `CN=openshift-signer@1518732226', issuer `CN=openshift-signer@1518732226', serial 0x01, RSA key 2048 bits, signed using RSA-SHA256, activated `2018-02-15 22:03:46 UTC', expires `2023-02-14 22:03:47 UTC', pin-sha256="aaNwFokUQMzgONAJJKb7rlabh/T7DC2qS6I/wsGK680="
- Status: The certificate is trusted. 
- Successfully sent 1 certificate(s) to server.
- Description: (TLS1.2)-(ECDHE-RSA-SECP256R1)-(CHACHA20-POLY1305)
- Session ID: 9E:FD:76:09:6F:A7:C7:1D:5A:C4:23:3C:A0:6F:5A:1D:C8:24:F7:19:C7:D2:0C:6E:4A:75:98:8B:9E:6D:0B:F1
- Ephemeral EC Diffie-Hellman parameters
 - Using curve: SECP256R1
 - Curve size: 256 bits
- Version: TLS1.2
- Key Exchange: ECDHE-RSA
- Server Signature: RSA-SHA256
- Client Signature: RSA-SHA256
- Cipher: CHACHA20-POLY1305
- MAC: AEAD
- Compression: NULL
- Options: safe renegotiation,
- Handshake was completed

- Simple Client Mode:

HTTP/1.0 401 Unauthorized
Cache-Control: no-store
Content-Type: application/json
Date: Thu, 22 Feb 2018 09:16:50 GMT
Content-Length: 129

{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Unauthorized","reason":"Unauthorized","code":401}
- Peer has closed the GnuTLS connection
```

**Comment by martinpitt** [comment]
I filed https://github.com/openshift/origin/issues/18715 about this.

**Comment by martinpitt** [comment]
This was [fixed in OpenShift 3.9](https://github.com/openshift/origin/pull/18837).
